### PR TITLE
[Flang] Add a HLFIR Minloc intrinsic

### PR DESCRIPTION
**Pull request by davemgreen**
The adds a hlfir minloc intrinsic, similar to the minval intrinsic already added, to help in the lowering of minloc. The idea is to later add maxloc too, and from there add a simplification for producing minloc with inlined elemental and hopefully less temporaries.

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
Is the SimplifyIntrinsics pass not working anymore for Minloc? (https://github.com/llvm/llvm-project/blob/383e35048e16c85ab26bc46822d3ceb11fd4d3f2/flang/lib/Optimizer/Transforms/SimplifyIntrinsics.cpp#L1370)

Eitherway, simplifying while lowering from HLFIR Ops is probably the right direction.

File: flang/lib/Lower/HlfirIntrinsics.cpp
Comment (jeanPerier) [reply flang/lib/Lower/HlfirIntrinsics.cpp:222]: Optional value need to use something like genOptionalValue from ConvertCall.cpp, assuming that the fallback value is zero/false.

**Review by jeanPerier — COMMENTED** [review]
Thank you for working on this

**Comment by davemgreen** [comment]
Thanks for the review.

File: flang/lib/Optimizer/HLFIR/IR/HLFIROps.cpp
Comment (davemgreen) [reply flang/lib/Optimizer/HLFIR/IR/HLFIROps.cpp:914]: It originally came from there, but is a bit different due to the result type not matching the input and often having a different shape than the reductions would.

File: flang/lib/Lower/HlfirIntrinsics.cpp
Comment (davemgreen) [reply flang/lib/Lower/HlfirIntrinsics.cpp]: Ah, that's useful. Thanks. I've added a new method for conditionally loading the value. Let me know if it seems OK.

File: flang/include/flang/Optimizer/HLFIR/HLFIROps.td
Comment (davemgreen) [reply flang/include/flang/Optimizer/HLFIR/HLFIROps.td]: I had considered that, but only went the other way as I noticed CountOp had a kind parameter already. I've had a go at changing it over to just use the return type.

**Review by tblah — APPROVED** [review]
This looks great to me! Thanks for your work.

Please wait for Jean's approval as well.

File: flang/lib/Optimizer/HLFIR/IR/HLFIROps.cpp
Comment (jeanPerier) [reply flang/lib/Optimizer/HLFIR/IR/HLFIROps.cpp:914]: OK, it looks the mask verification could still be shared into some verifyReductionMask helper. Custom result verification makes sense.

File: flang/include/flang/Optimizer/HLFIR/HLFIROps.td
Comment (jeanPerier) [reply flang/include/flang/Optimizer/HLFIR/HLFIROps.td]: Good point, I think we should remove it from CountOp afterwards then.

File: flang/lib/Lower/HlfirIntrinsics.cpp
Comment (jeanPerier) [reply flang/lib/Lower/HlfirIntrinsics.cpp]: Looks good, thanks.

**Review by jeanPerier — APPROVED** [review]
Looks great, thanks for addressing the comments!

**Review by jeanPerier — APPROVED** [review]
LGTM

**Pull request merged**
